File: android/src/main/kotlin/io/sariska/sariska_media_transport/ConferenceManager.kt
package io.sariska.sariska_media_transport

import android.os.Bundle
import com.facebook.react.bridge.ReadableMap
import io.sariska.sdk.Conference
import io.sariska.sdk.JitsiLocalTrack

class ConferenceManager(
        private val emit: (action: String, m: Map<String, Any?>?) -> Unit
) : Conference() {

    lateinit var conference: Conference

    fun createConference() {
        conference = Conference()
    }

    override fun join() {
        conference.join()
    }

    fun join(params: Map<String, *>) {
        conference.join(params["password"] as? String)
    }

    fun grantOwner(params: Map<String, *>) {
        conference.grantOwner(params["id"] as? String)
    }

    fun setStartMutedPolicy(params: Map<String, *>) {
        conference.setStartMutedPolicy(params["policy"] as? Bundle)
    }

    fun setReceiverVideoConstraint(params: Map<String, *>) {
        conference.setReceiverVideoConstraint(params["number"] as Int)
    }

    fun setSenderVideoConstraint(params: Map<String, *>) {
        conference.setSenderVideoConstraint(params["number"] as Int)
    }

    fun sendMessage(params: Map<String, *>) {
        if ((params["to"] as? String) != "") {
            conference.sendMessage(params["message"] as? String, params["to"] as? String)
        } else {
            conference.sendMessage(params["message"] as? String)
        }
    }

    fun setLastN(params: Map<String, *>) {
        conference.setLastN(params["num"] as Int)
    }

    fun dial(params: Map<String, *>) {
        conference.dial(params["number"] as Int)
    }

    fun muteParticipant(params: Map<String, *>) {
        if (params["mediaType"] as? String != "") {
            conference.muteParticipant(params["id"] as? String, params["mediaType"] as? String)
        } else {
            conference.muteParticipant(params["id"] as? String)
        }
    }

    fun setDisplayName(params: Map<String, *>) {
        conference.setDisplayName(params["displayName"] as? String)
    }

    fun addTrack(params: Map<String, *>) {
        for (track in conference.getLocalTracks()) {
            if (track.getId() == (params["trackId"] as? String)) {
                conference.addTrack(track)
            }
        }
    }

    fun removeTrack(params: Map<String, *>) {
        for (track in conference.getLocalTracks()) {
            if (track.getId() == (params["trackId"] as? String)) {
                conference.removeTrack(track)
            }
        }
    }

    fun replaceTrack(params: Map<String, *>) {
        var oldTrack : JitsiLocalTrack? = null
        var newTrack : JitsiLocalTrack? = null
        for (track in conference.getLocalTracks()) {
            if (track.getId() == (params["oldTrackId"] as? String)) {
                oldTrack = track
            }
            if (track.getId() == (params["newTrackId"] as? String)) {
                newTrack = track
            }
        }
        conference.replaceTrack(oldTrack, newTrack)
    }

    fun lock(params: Map<String, *>) {
        conference.lock(params["password"] as? String)
    }

    fun setSubject(params: Map<String, *>) {
        conference.lock(params["subject"] as? String)
    }

    override fun unlock() {
        conference.unlock()
    }

    fun kickParticipant(params: Map<String, *>) {
        conference.kickParticipant(params["id"] as? String)
    }

    fun pinParticipant(params: Map<String, *>) {
        conference.pinParticipant(params["id"] as? String)
    }

    fun selectParticipant(params: Map<String, *>) {
        conference.selectParticipant(params["id"] as? String)
    }

    fun selectParticipants(params: Map<String, *>) {
        conference.selectParticipants(params["participantIds"] as? Array<String>)
    }

    override fun startTranscriber() {
        conference.startTranscriber()
    }

    override fun stopTranscriber() {
        conference.stopTranscriber()
    }

    fun revokeOwner(params: Map<String, *>) {
        conference.revokeOwner(params["id"] as? String)
    }

    fun startRecording(params: Map<String, *>) {
        conference.startRecording(params["options"] as? Bundle)
    }

    fun stopRecording(params: Map<String, *>) {
        conference.stopRecording(params["sessionID"] as? String)
    }

    fun setLocalParticipantProperty(params: Map<String, *>) {
        conference.setLocalParticipantProperty(params["propertyKey"] as? String, params["propertyValue"] as? String)
    }

    fun removeLocalParticipantProperty(params: Map<String, *>) {
        conference.removeLocalParticipantProperty(params["name"] as? String)
    }

    fun sendFeedback(params: Map<String, *>) {
        conference.sendFeedback(params["overallFeedback"] as? String, params["detailedFeedback"] as? String)
    }

    override fun leave() {
        conference.leave()
    }

    fun destroy() {
        conference.leave()
    }

    fun release() {
       conference.leave()
    }

    override fun newConferenceMessage(action: String?, m: ReadableMap?) {
        if (action != null) {
            emit(action, m?.let { Utils.toMap(it) })
        }
    }
}

File: ios/Classes/ConferenceManager.swift
import Foundation
import sariska

class ConferenceManager: NSObject, Conference {

    private var emitter: (_ action: String, _ m: [String: Any?]?) -> Void
	let conference: Conference?

    init(_ emitter: @escaping (_ action: String, _ m: [String: Any?]?) -> Void) {
        self.emitter = emitter
    }

    func Release() {
        conference = nil
    }

    func createConference() {
        conference = Conference()
    }

    @objc func join() {
       conference.join()
    }

    @objc func join(_ params: NSDictionary) {
       conference.join(params["password"] as! NSString)
    }

    @objc func grantOwner(_ params: NSDictionary) {
       conference.grantOwner(params["id"] as! NSString)
    }

    @objc func setStartMutedPolicy(_ params: NSDictionary) {
       conference.setStartMutedPolicy(params["policy"] as! NSDictionary)
    }

    @objc func setReceiverVideoConstraint(_ params: NSDictionary) {
       conference.setReceiverVideoConstraint(params["number"] as! NSNumber)
    }

    @objc func setSenderVideoConstraint  (_ params: NSDictionary) {
       conference.setSenderVideoConstraint(params["number"] as! NSNumber)
    }

    @objc func sendMessage(_ params: NSDictionary) {
       if ((params["to"] as! NSString) != "") {
           conference.sendMessage(params["message"] as! NSString, params["to"] as! NSString)
       } else {
           conference.sendMessage(params["message"] as! NSString)
       }
    }

    @objc func setLastN(_ params: NSDictionary) {
       conference.setLastN(params["num"] as! NSNumber)
    }

    @objc func dial(_ params: NSDictionary) {
       conference.dial(params["number"] as! NSNumber)
    }

    @objc func muteParticipant(_ params: NSDictionary) {
       if (params["mediaType"] as! NSString) {
           conference.muteParticipant(params["id"] as! NSString, params["mediaType"] as! NSString)
       } else {
           conference.muteParticipant(params["id"] as! NSNumber)
       }
    }

    @objc func setDisplayName(_ params: NSDictionary) {
       conference.setDisplayName(params["displayName"] as! NSString)
    }

    @objc func addTrack(_ params: NSDictionary) {
        for track in conference.getLocalTracks() {
           let track = track as! JitsiLocalTrack
           if (track.getId() == (params["trackId"] as! NSString))  {
               conference.addTrack(track)
           }
       }
    }

    @objc func removeTrack(_ params: NSDictionary) {
        for track in conference.getLocalTracks() {
           let track = track as! JitsiLocalTrack
           if (track.getId() == (params["trackId"] as! NSString))  {
               conference.removeTrack(track)
           }
       }
    }

    @objc func replaceTrack(_ params: NSDictionary) {
        let oldTrack: JitsiLocalTrack?
        let newTrack: JitsiLocalTrack?
        for track in conference.getLocalTracks() {
           let track = track as! JitsiLocalTrack
           if (track.getId() == (params["oldTrackId"] as! NSString))  {
               oldTrack = track
           }
           if (track.getId() == (params["newTrackId"] as! NSString))  {
              newTrack = track
           }
       }
       conference.replaceTrack(oldTrack, newTrack)
    }

    @objc func lock(_ params: NSDictionary) {
       conference.lock(params["password"] as! NSString)
    }

    @objc func setSubject(_ params: NSDictionary) {
       conference.lock(params["subject"] as! NSString)
    }

    @objc func unlock() {
       conference.unlock()
    }

    @objc func kickParticipant(_ params: NSDictionary) {
       conference.kickParticipant(params["id"] as! NSString)
    }

    @objc func pinParticipant(_ params: NSDictionary) {
       conference.pinParticipant(params["id"] as! NSString)
    }

    @objc func selectParticipant(_ params: NSDictionary) {
       conference.selectParticipant(params["id"] as! NSString)
    }

    @objc func selectParticipants(_ params: NSDictionary) {
       conference.selectParticipants(params["participantIds"] as! NSArray)
    }

    @objc func startTranscriber() {
       conference.startTranscriber()
    }

    @objc func stopTranscriber() {
       conference.stopTranscriber()
    }

    @objc revokeOwner func (_ params: NSDictionary) {
       conference.revokeOwner(params["id"] as! NSString)
    }

    @objc func startRecording(_ params: NSDictionary) {
       conference.startRecording(params["options"] as! NSDictionary)
    }

    @objc func stopRecording(_ params: NSDictionary) {
       conference.stopRecording(params["sessionID"] as! NSString)
    }

    @objc func setLocalParticipantProperty(_ params: NSDictionary) {
       conference.setLocalParticipantProperty(params["propertyKey"] as! NSString, params["propertyValue"] as! NSString)
    }

    @objc func removeLocalParticipantProperty(_ params: NSDictionary) {
       conference.removeLocalParticipantProperty(params["name"] as! NSString)
    }

    @objc func sendFeedback(_ params: NSDictionary) {
       conference.sendFeedback(params["overallFeedback"] as! NSString, params["detailedFeedback"] as! NSString)
    }

    @objc func leave() {
       conference.leave()
    }

    @objc func destroy() {
        self?.Release()
    }

    @objc func newConferenceMessage(_ action: NSString, _ m: NSDictionary) {
       emitter(action, m)
    }
}